### PR TITLE
Musa support native fused logp kernel

### PR DESCRIPTION
**Pull request by Arlo-mt**
# [MUSA][kernels] Add native MUSA fused_logp kernel

## Summary

Adds a native MUSA implementation for the generic `fused_logp` kernel.

The MUSA implementation computes selected-token log probabilities and their
input gradients directly from logits using row-wise fused max and sum-exp
reductions. It supports FP32, FP16, and BF16 inputs, preserves the existing
`FusedLogpGenericOp` autograd wrapper, and falls back to the existing PyTorch
implementation when the MUSA extension is unavailable.

| Path | Status |
| --- | --- |
| MUSA forward (`csrc/musa/fused_logp_kernel.mu`) | Row-wise fused max/sum-exp kernel. New. |
| MUSA backward (`csrc/musa/fused_logp_kernel.mu`) | Row-wise native gradient kernel that avoids materializing the full softmax. New. |
| MUSA extension binding (`csrc/musa/ops.cpp`) | Validated `fused_logp` and `fused_logp_backward` entry points. New. |
| Python dispatch | MUSA `logp` selects the fused backend through the kernel registry. New. |
| Backward | MUSA uses the native backward kernel; other platforms retain the existing autograd path. |
| CUDA / ROCm / CPU behavior | Preserved. |
| Tests | Added MUSA forward, backward, and registry coverage. |
| Benchmark | MUSA S5000 results collected using the same shapes as the official CUDA fused-logp benchmark. |

## Implementation

- Added `csrc/musa/fused_logp_kernel.mu`.
  - Uses one CTA per logits row.
  - Computes the row maximum and sum-exp in FP32.
  - Gathers the selected-token logit directly from the input row.
  - Computes `grad_output * (one_hot(target) - softmax(logits))` in the native backward kernel.
  - Recomputes row statistics in the backward pass without allocating an `[N, V]` probability tensor.
  - Supports FP32, FP16, and BF16 logits.
  - Returns the output in the input dtype.
- Added `csrc/musa/ops.cpp`.
  - Exposes the MUSA `fused_logp` and `fused_logp_backward` bindings.
  - Validates device, shape, dtype, token range, and row count.
  - Converts input tensors to contiguous storage before launching the kernel.
- Updated `setup.py`.
  - Detects an available MUSA build environment.
  - Uses `MUSAExtension` for MUSA builds.
  - Compiles only the MUSA fused-logp sources in the MUSA path.
  - Preserves the existing CUDA and ROCm extension paths.
- Updated `rl_engine/kernels/registry.py`.
  - Adds the `MUSA_FUSED_LOGP_GENERIC` backend.
  - Routes the MUSA `logp` operation to `FusedLogpGenericOp`.
  - Keeps indexed, online, and deterministic log-probability variants on their existing backends.
- Added `tests/test_musa_fused_logp.py`.
  - Verifies native extension symbol availability.
  - Checks forward results against `log_softmax + gather`.
- Verifies native backward execution and finite input gradients.
  - Confirms MUSA registry dispatch.

## Validation environment

| Item | Value |
| --- | --- |
| GPU | Moore Threads MTT S5000 |
| GPU count | 8 |
| Benchmark devices | 1 |
| MUSA runtime | 40305 |
| Driver | 3.3.5-server |
| PyTorch | 2.7.1 |
| torch_musa | 2.7.1+5f0ecd1 |
| Host compiler | mcc 5.2.0 |
| MUSA architecture | `mp_31` |
| Python | 3.10 |

### Build

```text
RL_KERNEL_REQUIRE_EXT=1 \
python3 -m pip install --no-build-isolation --no-deps -e .

Result: successfully built and installed RL-Kernel
```

### MUSA-specific tests

```text
python3 -m pytest tests/test_musa_fused_logp.py -q

Result: 2 passed
```

The MUSA-specific tests cover:

- Native `_C.fused_logp` symbol availability.
- Forward correctness against `log_softmax + gather`.
- Autograd backward execution.
- Finite input gradients.
- MUSA registry selection.

### Existing fused-logp accuracy tests

```text
python3 -m pytest tests/test_op_accuracy.py \
  -q -k "fused_logp"

Result: 8 passed, 7 skipped, 3 deselected
```

### Registry tests

```text
python3 -m pytest tests/test_kernel_registry.py -q

Result: 12 passed
```

## Benchmarks

Single MTT S5000, one GPU, FP16, 5 warmup iterations and 20 measured iterations. The native and reference implementations both run on the same MUSA device. The reference implementation uses `log_softmax + gather`.

The benchmark shapes match the official CUDA fused-logp benchmark:

- `batch=16`, `seq_len=512`, `vocab=128256`
- `batch=32`, `seq_len=512`, `vocab=128256`

### Forward

| Shape (`Batch x Seq x Vocab`) | Dtype | Native MUSA (ms) | PyTorch reference (ms) | Native / reference | Tokens/s | Native peak VRAM | Reference peak VRAM |
| --- | --- | ---: | ---: | ---: | ---: | ---: | ---: |
| `16 x 512 x 128256` | `float16` | 9.303 | 14.669 | 1.58x faster | 880,583 | 1.96 GB | 9.79 GB |
| `32 x 512 x 128256` | `float16` | 18.036 | 29.439 | 1.63x faster | 908,415 | 3.91 GB | 19.57 GB |

The maximum absolute forward difference was `0` for both tested shapes after converting outputs to FP32 for comparison.

### Forward + Backward

The native path uses the MUSA `fused_logp` kernel for forward and the native MUSA `fused_logp_backward` kernel for backward.

| Shape (`Batch x Seq x Vocab`) | Dtype | Native fwd+bwd (ms) | PyTorch reference fwd+bwd (ms) | Native / reference | Native peak VRAM | Reference peak VRAM | Input grad max abs diff |
| --- | --- | ---: | ---: | ---: | ---: | ---: | ---: |
| `16 x 512 x 128256` | `float16` | 24.918 | 33.772 | 1.36x faster | 3.91 GB | 13.70 GB | 2.38e-7 |
| `32 x 512 x 128256` | `float16` | 48.811 | 67.729 | 1.39x faster | 7.83 GB | 27.40 GB | 4.77e-7 |

The native MUSA backward kernel removes the Python softmax materialization from the critical path. The end-to-end forward-plus-backward path is 1.36x faster for batch 16 and 1.39x faster for batch 32.

## Files

- `csrc/musa/fused_logp_kernel.mu`
- `csrc/musa/ops.cpp`
- `setup.py`
- `rl_engine/kernels/registry.py`
- `tests/test_musa_fused_logp.py`

## Limitations

- The native MUSA implementation covers the generic `fused_logp` forward and backward paths.
- Indexed, online, and deterministic log-probability variants continue to use their existing backends.
- The benchmark measures one MUSA S5000 device and should not be interpreted as a direct hardware comparison with the official H100 CUDA results.

## Summary by CodeRabbit

- **New Features**
  - Added MUSA-accelerated fused log-probability calculations for selected tokens.
  - Added forward and backward gradient computation for floating-point, half-precision, and bfloat16 tensors.
  - MUSA devices now automatically select the optimized fused log-probability implementation.
  - Added MUSA build support for the extension.

- **Bug Fixes**
  - Added validation for device, shape, dtype, and token-index compatibility.

- **Tests**
  - Added coverage for MUSA outputs, gradients, and backend selection.

### CODE REVIEW

**Comment by zhangj1an** [comment]
Thanks for your contribution!! 

I realise I can access MUSA computes via https://ai.gitee.com/compute/metax. I will try review this PR soon

**Review by Flink-ddd — APPROVED** [review]
LGTM
